### DISCUSSION TITLE
아이템 50. 적시에 방어적 본사본을 만들라

### DISCUSSION THREAD

**Discussion by chikeem90** [Study-2-Effective-Java/Effective-Java#126] [3. 과제]
- 클라이언트가 고의로든 고의가 아니든 악의적으로든 오작동하게 만들 수 있기 때문에 방어적으로 프로그래밍을 해야 함.
- Date의 경우 불변 객체가 아니기 때문에 생성 후 언제든 변경할 수 있음.
- set으로 열려있기 때문에 변경하게 되면 의도치 않은 결과를 초래할 수 있음.
- 이러한 문제를 예방하기 위해 Date가 아닌 Instant를 사용하거나 Period에서 방어적 복사본을 만들도록 구현 필요.

### 방어적 복사본 적용 예 - 1
```java
Date start = new Date();
Date end = new Date();
Period p = new Period(start, end);
end.setYear(78);
```

```java
public Period(Date start, Date end) {
    this.start = new Date(start.getTime());
    this.end = new Date(end.getTime());
}
```

### 방어적 복사본 적용 예 - 2
```java
public Date start() {
    return new Date(start.getTime());
}

public Date end() {
    return new Date(end.getTime());
}
```

### 결론
- 되도록 불변 객체들을 조합해 객체를 구성하는 것이 베스트
- 아닌 경우 방어적 복사본을 만드는 형태로 적용할 것

**Comment by Irisation23** [comment]
좀 더 이른 시점에 질문을 드렸어야 했는데 쪼꼼 늦었습니다.

https://d2.naver.com/helloworld/1250

의 글을 참조하다 보니 꽤 오래된 글임에도 불구하고 방어적 코드에 대한 작성이 있었더라고요.

``` java

protected void map(Text key, Text value, Context context) throws IOException, InterruptedException {

Matcher patternMatcher = regex.matcher(value.toString());


// 아파치 서버 웹 로그 패턴이 정의한 로그 패턴과 일치하는지 체크

if (patternMatcher.matches()) {

accessUrl = StringUtils.trimToEmpty(patternMatcher.group(2));    // client가 실제로 request 요청한 URL 부분

responseCode = StringUtils.trimToEmpty(patternMatcher.group(3));    // request에 대한 responset code 부분


// 요청 URL이 음식 End Page이고, HTTP 응답코드가 200(OK)일 경우

if (accessUrl.indexOf(FOOD_URL_PATTERN) > -1 && responseCode.startsWith("200")) {

    codeValue = StringUtils.substringBetween(accessUrl, CODE_BEFORE_TEXT, CODE_AFTER_TEXT);

    // 코드값(codeValue)이 없을 경우, codeValue는 null이 할당되므로 아래 trimToEmpty로 방어코드.

    foodId.set(StringUtils.trimToEmpty(codeValue));

    context.write(foodId, one);

}

}

}

```

```
    // 코드값(codeValue)이 없을 경우, codeValue는 null이 할당되므로 아래 trimToEmpty로 방어코드.

    foodId.set(StringUtils.trimToEmpty(codeValue));
```
라는 부분이 보였고 해당 `메서드(trimToEmpty)` 를 찾아보니 

![image](https://user-images.githubusercontent.com/53285909/219867051-945b1856-a3f0-49bf-9bd5-bc5c7697d70c.png)

StringUtils의 trimToEmpty 메서드는 문자열을 인자로 받아서, 해당 문자열이 null이거나 공백 문자열인 경우 빈 문자열("")을 반환하는 메서드입니다. 
이게 방어적 코드라는 얘기가 약간은 어려운 개념인것 같아 질문 드렸습니다.

어떤 이유 때문에 해당 코드가 방어적 코드로서의 동작을 하는것일까요?

**Reply by chikeem90** [reply]
@bunsung92 님
이제 확인했네요,, ㅎㅎ 아주 오래된 글을 ㅎㅎ
링크 주신 내용을 읽어보니 food Id가 없는 경우 codeValue가 null일텐데, 그 때 context.write() 메소드의 key가 null이 들어가게되어 에러가 발생하는 것을 막고자 trimToEmpty를 사용한 거 같습니다. (저도 map reduce의 정확한 동작방식은 몰라서,, 추측,,)
정리하자면 NPE가 발생할 수 있는 가능성을 막기 위해 trimToEmpty를 사용하였고, 결과적으로 null이어도 빈 문자열로 처리되게끔 하여 로직이 수행되는 동안 에러가 발생하지 않도록,,?

**Reply by Irisation23** [reply]
답변 감사합니다. 저희 스터디원들은 새벽이 활발하네요 👍 

방어적 로직을 구현한 기술 블로그가 있지 않을까 해서 좀 뒤져보다보니 신나보이는 코드가 있어서 슥 올려봤습니다.
말씀하신 내용을 토대로 생각해 보면 `NPE` 를 방지하는 행위 자체가 방어적 코드라고 생각하면 될까요??

**Reply by chikeem90** [reply]
@bunsung92 님
어,, 'NPE를 방지하는 행위 = 방어적 코드'인건 아니지만 방어적 코드 중에 하나인 거라고 보시면 될 거 같아요 
책에서도 나왔듯이 클라이언트는 우리가 의도하지 않은 값을 언제든 입력할 수도 있기 때문에, 우리가 개발한 로직이 문제 없이 (의도치 않은 에러로 중단되거나 쓰레기값이 저장되어 잘못 처리된다거나) 잘 돌아가게 하기 위해서는 에러가 발생하는 케이스를 미리 예측해서 처리해두는 것을 방어적 코드라고 이해하시면 될 거 같네요

**Reply by Irisation23** [reply]
- NPE가 하나의 행위라는 것을 말하고 싶었는데 제 질문이 모호했던것 같네요. 😢

@chikeem90 님이 답변해 주신 내용을 토대로 `방어적 프로그래밍`에 관해 글을 몇개 읽었는데 이해하기 정말 좋았습니다. 

![image](https://user-images.githubusercontent.com/53285909/220235183-e7af91a9-8b66-4ee7-8a18-47777d26039a.png)

제가 했던 질문은 위의 정리에서 1. 상황에서의 방어적 코드였다고 생각해 볼 수있겠네요. 궁금증이 싹 풀렸습니다!

> 내용 출처
> https://a-road-after-walking.tistory.com/54

**Comment by corock** [comment]
내용 잘 읽었습니다~

하나 궁금한 게, 이전에 시니어 개발자 중 한 분에게 가변 객체와 불변 객체가 있을 때 무조건 불변 객체가 좋나요? 라고 질문한 적이 있습니다.
그때 받은 답변으로, 되게 조심스러웠지만 "그렇다." 였는데요.

저도 무지성 setter 사용을 지양하고 불변 객체를 선호하지만, 한편으로는 객체를 생성하는 비용을 생각하지 않을 수 없습니다.
이것에 대해서 어떻게 생각하시는지 궁금하고, 현재 팀 내에서는 어떻게 운용하고 있는지 알려주실 수 있나요~?

**Reply by JoisFe** [reply]
불변 객체를 만드는 것이 객체 생성 비용을 높이지만 오히려 성능상의 이점을 가져올 수 있다고 하네요
불변 객체 경우 thread safe 하기도 하고 오히려 객체를 복사해야하는 데 필요한 비용이 줄어들기 때문이라 합니다.
그러나 객체 상태가 자주 변경된다면 불변 객체를 만드는 것이 적적하지 않다고 합니다.
어디서든 완벽한 정답은 없는 것 같습니다.!!

**Reply by Irisation23** [reply]
https://min103ju.github.io/effective%20java/immutable/
